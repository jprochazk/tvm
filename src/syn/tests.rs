use super::Parser;

fn _parse(input: &str) -> String {
  match Parser::new(input).try_parse() {
    Ok(ast) => format!("{ast:#?}"),
    Err(e) => e
      .into_iter()
      .map(|e| format!("{}", e.with_src(input)))
      .collect::<Vec<_>>()
      .join("\n"),
  }
}

macro_rules! parse {
  ($input:literal) => {
    _parse(indoc::indoc!($input))
  };
}

macro_rules! test {
  ($name:ident, $input:literal) => {
    #[test]
    fn $name() {
      insta::assert_snapshot!(parse!($input))
    }
  };
}

test! {
  variables,
  r#"
    let v: int = 0;
  "#
}

test! {
  opt_type,
  r#"
    let opt: int? = 1;
  "#
}

test! {
  bin_add,
  r#"
    2 + 2
  "#
}

test! {
  function,
  r#"
    fn f() {}
  "#
}

test! {
  lone_fn_keyword,
  r#"
    fn
  "#
}

test! {
  block_expr,
  r#"
    let v = do { 1 };
  "#
}
