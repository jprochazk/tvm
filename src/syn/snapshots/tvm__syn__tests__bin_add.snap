---
source: src/syn/tests.rs
expression: "parse!(r#\"\n    2 + 2\n  \"#)"
---
SyntaxTree {
    top_level: Block {
        span: Span {
            start: 0,
            end: 0,
        },
        body: [],
        tail: Some(
            BinaryExpr {
                left: LiteralExpr {
                    value: Int(
                        2,
                    ),
                },
                op: Add,
                right: LiteralExpr {
                    value: Int(
                        2,
                    ),
                },
            },
        ),
    },
}
